File: lib/data/api-response.ts
import { NextResponse } from "next/server";
import type { DataBackend } from "./contracts.ts";

export function dataResponse<T>(data: T, backend: DataBackend) {
  return NextResponse.json(
    {
      data,
      meta: {
        backend,
        generatedAt: new Date().toISOString(),
      },
    },
    { headers: { "Cache-Control": "no-store" } },
  );
}

export function dataServiceError(error: unknown) {
  console.error("SafeGo data service error", error);

  return NextResponse.json(
    {
      error: {
        code: "DATA_SERVICE_UNAVAILABLE",
        message: "SafeGo data is temporarily unavailable.",
      },
    },
    { status: 503, headers: { "Cache-Control": "no-store" } },
  );
}

File: server/demo/src/main/java/com/safego/demo/api/ApiResponse.java
package com.safego.demo.api;

import java.time.Instant;
import java.util.Map;

public final class ApiResponse {

    private ApiResponse() {}

    public static Map<String, Object> ok(Object data, String backend) {
        return Map.of(
            "data", data,
            "meta", Map.of(
                "backend", backend,
                "generatedAt", Instant.now().toString()
            )
        );
    }

    public static Map<String, Object> error(String code, String message) {
        return Map.of("error", Map.of("code", code, "message", message));
    }
}
